src/nifreeze/model/base.py

```diff
@@ -78,114 +79,81 @@ class BaseModel:
 
     """
 
-    __slots__ = (
-        "_model",
-        "_mask",
-        "_models",
-        "_datashape",
-        "_is_fitted",
-        "_modelargs",
-    )
+    __slots__ = {
+        "_dataset": "Reference to a :obj:`~nifreeze.data.base.BaseDataset` object.",
+    }
 
-    def __init__(self, mask=None, **kwargs):
+    def __init__(self, dataset, **kwargs):
         """Base initialization."""
 
-        # Keep model state
-        self._model = None  # "Main" model
-        self._models = None  # For parallel (chunked) execution
-
-        # Setup brain mask
-        if mask is None:
+        self._dataset = dataset
+        # Warn if mask not present
+        if dataset.brainmask is None:
             warn(
                 "No mask provided; consider using a mask to avoid issues in model optimization.",
                 stacklevel=2,
             )
 
-        self._mask = mask
-
-        self._datashape = None
-        self._is_fitted = False
-
-        self._modelargs = ()
-
-    @property
-    def is_fitted(self):
-        return self._is_fitted
-
-    def fit(self, data, **kwargs):
-        """Abstract member signature of fit()."""
-        raise NotImplementedError("Cannot call fit() on a BaseModel instance.")
-
-    def predict(self, *args, **kwargs):
-        """Abstract member signature of predict()."""
-        raise NotImplementedError("Cannot call predict() on a BaseModel instance.")
+    def fit_predict(self, *_, **kwargs):
```

Comment:
We don't need to expose a fit/predict interface here. We'll fit&predict all the time. For stationary models (meaning, those trained on the full dataset, e.g., to fine tune hyperparameters such as GPR, will need to implement this logic)